Update test suite, update Solidity version for compiling test contracts, remove deprecated JSON-RPC methods

fselmo opened the pull request:
### What was wrong?

This PR was meant to update Solidity contract versions but it revealed that `geth --dev` was broken. Geth fixed this and this PR brings those changes in (`v1.14.3`). Updated geth no longer has support for `eth_mining`, `eth_hashrate`, `eth_submitWork`, `eth_getWork`, `eth_coinbase`. This PR gets our test suite updated with all of these changes as well as removes support for those methods altogether. We will have to issue a backport PR for `v6` to deprecate all of these.

- update py-geth
- update geth version to `v1.14.3`
- update `eth-tester[py-evm]` to `0.10.0b.4`
- closes #3405 
- remove support for `eth_mining`, `eth_hashrate`, `eth_submitHashrate`, `eth_submitWork`, `eth_getWork`, `eth_coinbase`
- remove support for `pending` block

---

The goal here was to compile test contracts with Solidity ``0.8.25`` which broke because we needed to update geth. This PR was open for a while and `0.8.26` came out as well. [This build](https://app.circleci.com/pipelines/github/ethereum/web3.py/6756) has everything working with `0.8.25` and then the latest version should use `0.8.26`.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

`0_o`

fselmo committed:
Update test fixture to use Sol ``v0.8.25``; Update geth:

- Remove pending block reliance; set dev period to 1 second
- Use the latest geth 1.14.5 for integration; re-generate fixture
- Tweaks to the fixture generation based on geth updates
- re-compile test contracts with Solidity v0.8.25
- Remove support for non-existent client properties:
  - ``eth_coinbase``
  - ``eth_mining``
  - ``eth_hashrate``
  - ``eth_submitHashrate``
  - ``eth_submitWork``
  - ``eth_getWork``
- Update py-geth to use the newly released beta
- quick lint fixes
- Fix remaining tests that rely on mining:
- newsfragment for #3307

fselmo commented:
fyi: I updated my gpg key so I had to squash all old commits into one large commit that was signed by the new gpg key. I would've used a bit more separation of concerns in the commits / messages otherwise.

fselmo committed:
Compile test contracts with Solidity v0.8.26

fselmo committed:
Give more stability to tests using either "pending" or "latest":

- Set a timeout and loop over the latest block a few times until desired
  tx is included.

- Some processing, such as async IPC, goes faster than others. There'stability
  not a great way to ensure if we are in the "pending" or "latest" state for
  blocks using geth --dev. The important thing for testing on our end is that the
  JSON-RPC method works the way it's expected. Whether or not it returns the
  expected result based on block number, etc, is on Geth's side to make sure.
  Test the JSON-RPC method and give it flexibility in this commit.

reedsa reviewed (commented):
Looks good so far. Not sure what happened with the latest run in CI but I'll have another look with the next `py-geth` release.

fselmo commented:
> Not sure what happened with the latest run in CI

Circle CI was down EOD yesterday. I meant to re-run today. 

> I'll have another look with the next `py-geth` release

I don't think we plan on releasing it while keeping this PR open. This is going in *web3.py* `main` which is also in beta so I think we should merge it as is and just make sure to update to stable before releasing v7 stable.

fselmo committed:
Simplify old contract state test

kclowes commented on tests/integration/generate_fixtures/go_ethereum.py:
```diff
@@ -104,8 +100,8 @@ def get_geth_process(geth_binary, datadir, geth_port):
         stdin=subprocess.PIPE,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.PIPE,
+        # stdout=subprocess.PIPE,
```
Should these go? Or be uncommented?

fselmo replied:
Yes this was in order to get better logs, thank you!

kclowes commented on tests/integration/generate_fixtures/go_ethereum.py:
```diff
@@ -115,8 +111,8 @@ def get_geth_process(geth_binary, datadir, geth_port):
         "Geth Process Exited:\n"
-        f"stdout:{to_text(output)}\n\n"
-        f"stderr:{to_text(errors)}\n\n"
+        # f"stdout:{to_text(output)}\n\n"
```
Same here - can these go or should they be put back?

kclowes reviewed (approved):
LGTM! I left a few inline nits, and then the only other thing is I wonder if we should think of a better name for `accounts[0]` to reference it by. Maybe `default_account`? Feel free to take or leave, I don't think it's a huge improvement and I don't really have a better idea at the moment ¯\_(ツ)_/¯

fselmo commented:
> LGTM! I left a few inline nits, and then the only other thing is I wonder if we should think of a better name for `accounts[0]` to reference it by. Maybe `default_account`? Feel free to take or leave, I don't think it's a huge improvement and I don't really have a better idea at the moment ¯_(ツ)_/¯

Yeah, agreed. I think that might be a nice pytest fixture abstraction to have if anything. I'll take a peek while I'm in there.

fselmo committed:
Use ``eth.default_account`` for tests, in the age of no ``eth_coinbase``

This commit updates the tests to set ``w3.eth.default_account`` to the
first account and use it in place of ``w3.eth.coinbase``, since
``eth_coinbase`` was removed in clients like geth. Using
``w3.eth.accounts[0]`` is not a very user-friendly API to keep calling
across all tests.

fselmo commented:
@kclowes, I opted to make use of the `w3.eth.default_account`, [in the most recent commit](https://github.com/ethereum/web3.py/pull/3307/commits/46a717a63b88bd19a0e249bf4546f31f7a17849b), by setting that to the first account for tests. I think that makes the most sense, does that feel like a good resolution to you? I added a test that makes sure `w3.eth.default_account` is empty by default since that was being checked in another test.

kclowes reviewed (approved):
Yeah, I like that! 🚢
